### PR TITLE
Add suggestions for std_instead_of_core

### LINKED ISSUE #11446: Add suggestions for `std_instead_of_core`
### Description

If the `std_instead_of_core` lint (and its relatives `std_instead_of_alloc` and `alloc_instead_of_core`) generated suggested fixes, then it would be easy to bulk migrate imports using `cargo clippy --fix`, which would make it easier to explore adding `no_std` compatibility to a library.

### Version

_No response_

### Additional Labels

@rustbot label +C-enhancement

### PR DESCRIPTION
**Pull request by tom-anders**
```
changelog: [`std_instead_of_core`]: add suggestions
```

Fixes #11446

### CODE REVIEW

**Comment by Jarcho** [comment]
This needs tests (and an implementation, or at least no suggestion) for the more fun cases with mixed imports. e.g.
```rust
use std::fmt::{Result, format};
use std::{collections::HashMap, fs, str};
```
Should end up as
```rust
use core::fmt::Result;
use alloc::fmt::format;
use alloc::collections::HashMap;
use std::fs;
use core::str;
``

File: tests/ui/std_instead_of_core.stderr [hunk 1]
```diff
@@ -2,100 +2,73 @@ error: used import from `std` instead of `core`
...
 
 error: used import from `std` instead of `core`
-  --> $DIR/std_instead_of_core.rs:12:9
+  --> $DIR/std_instead_of_core.rs:12:11
    |
 LL |     use ::std::hash::Hash;
-   |         ^^^^^^^^^^^^^^^^^
-   |
-   = help: consider importing the item from `core`
+   |           ^^^ help: consider importing the item from `core`: `core`
```
Comment (Manishearth) [inline tests/ui/std_instead_of_core.stderr:13]: hm, not a fan of how this looks where "core" gets repeated. it sounds confusing but it makes sense if you understand how the span arrows work.

thoughts on better ways to present this? perhaps "consider importing the item using: `core`"?

unsure if that's better, open to opinions!

**Review by Manishearth — COMMENTED** [review]
overall looks good but i want to discuss the exact text a bit, and i agree with @jarcho that more examples would be great!

**Comment by cathaysia** [comment]
Is this turned off by default? I mean, will this have any impact on no_std programs. Or is it automatically disabled if no_std is found?

**Comment by Manishearth** [comment]
No, these are restriction lints

(and this PR is not adding them, it is improving the existing ones. We've had them for a while)

**Comment by tom-anders** [comment]
> This needs tests for the more fun cases with mixed imports. e.g.
> 
> ```rust
> use std::fmt::{Result, format};
> use std::{collections::HashMap, fs, str};
> ```
> 
> Should end up as (if there's going to be a suggestion for them)
> 
> ```rust
> use core::fmt::Result;
> use alloc::fmt::format;
> use alloc::collections::HashMap;
> use std::fs;
> use core::str;
> ```
> 
> I don't remember how those get picked up by the lint.

Currently this fails with `failed to apply suggestions for tests/ui/std_instead_of_core.rs with rustfix: Cannot replace slice of data that was already replaced`.

Getting the replacement right here seems complicated - If you can give me some starting points on how to handle this, I can have a go though.

Otherwise, maybe let's just detect this case and not offer any suggestions for now? We could open a new issue for handling these kinds of replacements.

**Comment by Manishearth** [comment]
Yeah that sounds fine. We should also file an issue on rustfix for that.

(ideally take the JSON-format rust output and the file and include them in the issue)

**Comment by tom-anders** [comment]
> Yeah that sounds fine. We should also file an issue on rustfix for that.
> 
> (ideally take the JSON-format rust output and the file and include them in the issue)

Is it really a rustfix issue though? Currently our lint will first try to replace "std" with with core (for Result import) and then the same "std" with alloc (for the format import). Do you think rustfix should simply apply the second suggestion in this case?

**Pull request merged**